source/client.c

```diff
@@ -598,90 +716,18 @@ struct aws_mqtt_client_connection *aws_mqtt_client_connection_new(struct aws_mqt
-                (void *)connection);
-            mqtt_connection_unlock_synced_data(connection);
-            return;
-        }
 
-        mqtt_connection_unlock_synced_data(connection);
-    } /* END CRITICAL SECTION */
-    /* If the slot is not NULL, the connection is still connected, which should be prevented from calling this function
-     */
-    AWS_ASSERT(!connection->slot);
-    /* clean up the pending_requests if it's not empty */
-    while (!aws_linked_list_empty(&connection->synced_data.pending_requests_list)) {
-        struct aws_linked_list_node *node = aws_linked_list_pop_front(&connection->synced_data.pending_requests_list);
-        struct aws_mqtt_outstanding_request *request =
-            AWS_CONTAINER_OF(node, struct aws_mqtt_outstanding_request, list_node);
-        /* Fire the callback and clean up the memory, as the connection get destoried. */
-        if (!request->completed) {
-            request->on_complete(
-                request->connection, request->packet_id, AWS_ERROR_MQTT_CONNECTION_DESTROYED, request->on_complete_ud);
-        }
-        aws_memory_pool_release(&connection->synced_data.requests_pool, request);
-    }
-
-    AWS_LOGF_DEBUG(AWS_LS_MQTT_CLIENT, "id=%p: Destroying connection", (void *)connection);
-
-    /* If the reconnect_task isn't freed, free it */
-    if (connection->reconnect_task) {
-        aws_mem_release(connection->reconnect_task->allocator, connection->reconnect_task);
-    }
-    aws_string_destroy(connection->host_name);
-
-    /* Clear the credentials */
-    if (connection->username) {
-        aws_string_destroy_secure(connection->username);
-    }
-    if (connection->password) {
-        aws_string_destroy_secure(connection->password);
-    }
-
-    /* Clean up the will */
-    aws_byte_buf_clean_up(&connection->will.topic);
-    aws_byte_buf_clean_up(&connection->will.payload);
-
-    /* Clear the client_id */
-    aws_byte_buf_clean_up(&connection->client_id);
-
-    /* Free all of the active subscriptions */
-    aws_mqtt_topic_tree_clean_up(&connection->thread_data.subscriptions);
-
-    aws_hash_table_clean_up(&connection->synced_data.outstanding_requests_table);
-    aws_memory_pool_clean_up(&connection->synced_data.requests_pool);
-
-    aws_mutex_clean_up(&connection->synced_data.lock);
-
-    aws_tls_connection_options_clean_up(&connection->tls_options);
-
-    /* Clean up the websocket proxy options */
-    if (connection->websocket.proxy) {
-        aws_tls_connection_options_clean_up(&connection->websocket.proxy->tls_options);
+    return connection;
+}
 
-        aws_mem_release(connection->allocator, connection->websocket.proxy);
-        connection->websocket.proxy = NULL;
-        connection->websocket.proxy_options = NULL;
+void aws_mqtt_client_connection_release(struct aws_mqtt_client_connection *connection) {
+    if (connection != NULL) {
+        aws_ref_count_release(&connection->ref_count);
```

Comment:
Just want to make sure, because the destroy of connection is asynced call. So it is possible to call `aws_ref_count_release` when the refcount is zero already but the connection is still not NULL, right? It will lead to a crash. Is that what we want?

Reply:
If we want it to crash for negative refcount, can we do `ASSERT(connection)` to make it more consistent?

Reply:
Well sure it's possible but that's 100% user error and not something I think we need to worry about.

Reply:
Would it be better to `ASSERT(connection)`, in which case, the invalid call off `aws_mqtt_client_connection_release ` will always lead to a crash. For now, it may crash or silently do nothing.

Reply:
All of the release functions check null, which makes it easier to write unconditional tear-down functions that work regardless of partial construction (i.e. got halfway through _new() and failed).

Reply:
aws_mqtt_connection_new has the old approach which has many goto-destinations, one for each thing that needs to be cleaned up and everything has to be perfectly aligned.  More recently, we just make more resilient _destroy or _release functions and call it regardless of how far we got in construction